packages/edge/src/response.ts

```diff
@@ -20,3 +20,51 @@ export function json(data: any, init?: ResponseInit): Response {
   // @ts-expect-error This is not in lib/dom right now, and we can't augment it.
   return Response.json(data, init);
 }
+
+/**
+ * Builds a response for returning data based on promise that take many seconds to resolve.
+ * The response is returned immediately, but data is only written to it when the promise resolves.
+ *
+ * @param dataPromise Promise for data to be sent as the response body. Note, that if this promise is
+ *     rejected, then a plain text "ERROR" is returned to the cliet. Catch errors on the promise yourself
+ *     to add custom error handling.
+ * @param init optional custom response status, statusText and headers
+ *
+ * @example
+ * ```ts
+ * import { potentiallyLongRunningResponse } from '@vercel/edge';
+ *
+ * export default () => {
+ *   const slowPromise = new Promise((resolve) => setTimeout(() => resolve("Done"), 20000));
+ *   return potentiallyLongRunningResponse(slowPromise);
+ * };
+ * ```
+ */
+export function potentiallyLongRunningResponse(
+  dataPromise: Promise<string | Uint8Array>,
+  init?: ResponseInit
+): Response {
+  return new Response(
+    new ReadableStream({
+      start(controller) {
+        dataPromise
+          .then((data: string | Uint8Array) => {
+            if (typeof data === 'string') {
+              controller.enqueue(new TextEncoder().encode(data));
+            } else {
+              controller.enqueue(data);
+            }
+            controller.close();
+          })
+          .catch(error => {
+            console.log(
+              `Error in 'potentiallyLongRunningResponse' dataPromise: ${error}`
+            );
+            controller.enqueue(new TextEncoder().encode('ERROR'));
```

Comment:
**suggestion:** shouldn't we return the error text instead of a fixed string?

Reply:
> If folks want something better they need to handle the error themselves

<picture data-single-emoji=":sgtm:" title=":sgtm:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..WEpCGl6oBK7j_6Bb.5blV7WMVAx8TBsAv9nEmqUPYTvFVjCpkUwssn_jnIKrrC4KLgbTCcOcOnLpE3jdcwrlyNy2DaSIiGWv1ZOeZl6eRbqEcNw2IvXxw604C-L7XDWucjE8.F3k5z-vIUrlJ-lyhTDos0g" alt=":sgtm:" align="absmiddle"></picture>

If only we had a typed error in Promse in TypeScript (`Promise<Ok, Err>`), we could ask for a `Promise<Ok, never>` in this function to explain this using types <picture data-single-emoji=":wen:" title=":wen:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..pE7JYhbIFMIhGBpv.og1oNamuaC-gzt_TtFr6svqReraaZ7mVkdiV_08Jhg7uOwvamGlF0lB8-mVFs5iCIRw8to84TQ8XhNsHvBfVTH19tSVApFAkLreRILoYSBpYUPXwiA.Y2zywBR9EN0b2kAtgNztqA" alt=":wen:" align="absmiddle"></picture> <picture data-single-emoji=":p0:" title=":p0:"><img class="emoji" width="20" height="auto" src="https://single-emoji.vercel.app/api/emoji/eyJhbGciOiJkaXIiLCJlbmMiOiJBMjU2R0NNIn0..IAgzjsZzIl4qgwC2.N_AYiA_To_KJUFpjPZM16dQqfybEB1HRP1UDdZK4y8l8j9NKi6FEmtS0CHig_csIQLCYCVy-HeGcSHgcn3lBLqDiBODc4fouJJNR1fMwG3TO60VV.bM-fIbQWE2iIWD_5EcS8Jg" alt=":p0:" align="absmiddle"></picture>